content/academy/expert_scraping_with_apify/solutions/rotating_proxies.md

```diff
@@ -0,0 +1,115 @@
+---
+title: VII - Rotating proxies/sessions
+description: Learn first hand how to rotate proxies and sessions in order to avoid the majority of the most common anti-scraping protections.
+menuWeight: 7
+paths:
+    - expert-scraping-with-apify/solutions/rotating-proxies
+---
+
+# [](#rotating-proxy-sessions) Rotating proxies/sessions
+
+If you take a look at our current code for the Amazon scraping actor, you might notice this snippet:
+
+```JavaScript
+const proxyConfiguration = await Apify.createProxyConfiguration({
+    groups: ['RESIDENTIAL'],
+});
+```
+
+We didn't provide much explanation for this initially, as it was not directly relevant to the lesson at hand. When you [create a **ProxyConfiguration**]({{@link anti_scraping/proxies/using_proxies.md}}) and pass it to a crawler, the SDK will make the crawler automatically rotate through the proxies. This entire time, we've been using the **RESIDENTIAL** proxy group to avoid being blocked by Amazon.
+
+> Go ahead and try commenting out the proxy configuration code then running the scraper. What happens?
+
+In order to rotate sessions, we must utilize the [**SessionPool**](https://sdk.apify.com/docs/guides/session-management), which we've actually also already been using by setting the **useSessionPool** option in our crawler's configuration to **true**. The SessionPool advances the concept of proxy rotation by tying proxies to user-like sessions and rotating those instead. In addition to a proxy, each user-like session has cookies attached to it (and potentially a browser-fingerprint as well).
+
+## [](#configuring-session-pool) Configuring SessionPool
+
+Let's go ahead and add a **sessionPoolOptions** key to our crawler's configuration so that we can modify the default settings:
+
+```JavaScript
+const crawler = new Apify.CheerioCrawler({
+    requestList,
+    requestQueue,
+    proxyConfiguration,
+    useSessionPool: true,
+    // This is where our session pool
+    // configuration lives
+    sessionPoolOptions: {
+        // We can add options for each
+        // session created by the session
+        // pool here
+        sessionOptions: {
+
+        },
+    },
+    maxConcurrency: 50,
+    // ...
+});
+```
+
+Now, we'll use the **maxUsageCount** key to force each session to be thrown away after 5 uses, and **maxErrorScore** to trash a session once it receives an error. Then, we'll use **persistStateKey** back in **sessionPoolOptions** to persist the session pool's state, therefore handling migrations:
+
+```JavaScript
+const crawler = new Apify.CheerioCrawler({
+    requestList,
+    requestQueue,
+    proxyConfiguration,
+    useSessionPool: true,
+    sessionPoolOptions: {
+        persistStateKey: 'AMAZON-SESSIONS',
```

Comment:
This is not needed to handle migrations, it stores automatically to local KV store. You can just choose the persist elsewhere. I would skip this in this section, it is super niche and parallel sessions don't work anyway